src/Ship.js

```diff
@@ -0,0 +1,29 @@
+function Ship(itinerary) {
+  this.itinerary = itinerary;
+  this.currentPort = itinerary.ports[0];
+  this.previousPort = null;
+  this.currentPort.addShip(this);
+
+  this.setSail = function () {
+    const itinerary = this.itinerary;
+    const currentPortIndex = itinerary.ports.indexOf(this.currentPort);
+
+    if (currentPortIndex === itinerary.ports.length - 1) {
+      throw new Error("End of itinerary reached");
+    }
+
+    this.previousPort = this.currentPort;
+    this.currentPort.removeShip(this);
+    this.currentPort = null;
+  };
+
+  this.dock = function (port) {
+    const itinerary = this.itinerary;
+    const previousPortIndex = itinerary.ports.indexOf(this.previousPort);
+
+    this.currentPort = itinerary.ports[previousPortIndex + 1];
```

Comment:
As above, could be simplified:

```suggestion
  this.dock = function(port) {
    const previousPortIndex = itinerary.ports.indexOf(this.previousPort);

    this.currentPort = this.itinerary.ports[
      this.itinerary.ports.indexOf(this.previousPort) + 1
    ];
```